src/main/scala/mesosphere/marathon/core/health/impl/AppHealthCheckActor.scala

```diff
@@ -0,0 +1,112 @@
+package mesosphere.marathon
+package core.health.impl
+
+import akka.actor.{ Actor, Props }
+import akka.event.EventStream
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.core.event.InstanceHealthChanged
+import mesosphere.marathon.core.health.impl.AppHealthCheckActor._
+import mesosphere.marathon.core.health.{ Health, HealthCheck }
+import mesosphere.marathon.core.instance.Instance
+import mesosphere.marathon.state.{ PathId, Timestamp }
+
+/**
+  * This actor aggregates the statuses of health checks at the application level
+  * in order to maintain a global healthiness status for each instance.
+  *
+  * @param eventBus The eventStream to publish status changed events to
+  */
+class AppHealthCheckActor(eventBus: EventStream) extends Actor with StrictLogging {
+  /**
+    * Map of health check definitions of all applications
+    */
+  private var healthChecks: Map[ApplicationKey, Set[HealthCheck]] = Map.empty
+
+  /**
+    *  Map of results of all health checks for all applications.
+    *  Results are optional, therefore the global health status of an instance
+    *  is either:
+    *    unknown (if some results are still missing)
+    *    healthy (if all results are known and healthy)
+    *    not healthy (if all results are known and at least one is unhealthy)
+    */
+  private var healthCheckStates: Map[InstanceKey, Map[HealthCheck, Option[Health]]] =
+    Map.empty
+
+
+  private def computeGlobalHealth(instanceHealthResults: Map[HealthCheck, Option[Health]]): Option[Boolean] = {
+    val instanceHealthResultsSet = instanceHealthResults.values.toSet
+    val healthy = instanceHealthResultsSet.forall(x => x.fold(false)(_.alive))
```

Comment:
Perhaps this is better?
```
instanceHealthResults.values.forall { health => health.alive.contains(true) }
```

Reply:
Indeed! I'll update it. Thanks.

Reply:
Actually it does not compile and for a good reason: health is the option and alive is a bool.